endpoint: signal when BPF program is compiled for the first time

ianvernon opened the pull request:
The Istio CI test has been unstable because of behavior in some pods where, if DNS fails to resolve, before the BPF program for the pod is installed, even after the pod has a BPF program installed, it will not be able to connect to the DNS server to resolve DNS again until the pod restarts.

Prior to this change, if Cilium is running in tandem with Istio, we returned before the BPF program was installed for the endpoint to avoid deadlock when configuring Envoy. To ensure that this sort of behavior in pods does not occur again, return from endpoint creation if the endpoint has been determined to be Istio-injected, and has a BPF program installed.

This PR adds a channel which is closed once the initial BPF compilation has occurred for an endpoint. The state of this channel is used to determine if a BPF program has been compiled for the endpoint.
    
Fixes: #5859
    
Signed-off by: Ian Vernon <ian@cilium.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5869)
<!-- Reviewable:end -->

Commit:
endpoint: signal when BPF program is compiled for the first time

The Istio CI test has been unstable because of behavior in some pods where, if
DNS fails to resolve, before the BPF program for the pod is installed, even
after the pod has a BPF program installed, it will not be able to connect to the
DNS server to resolve DNS again until the pod restarts.

Prior to this change, if Cilium is running in tandem with Istio, we returned
before the BPF program was installed for the endpoint to avoid deadlock when
configuring Envoy. To ensure that this sort of behavior in pods does not occur
again, return from endpoint creation if the endpoint has been determined to be
Istio-injected, and has a BPF program installed.

This PR adds a channel which is closed once the initial BPF compilation has
occurred for an endpoint. The state of this channel is used to determine if a
BPF program has been compiled for the endpoint.

Fixes: #5859

Signed-off by: Ian Vernon <ian@cilium.io>

tgraf commented on daemon/endpoint.go:
```diff
@@ -271,7 +271,7 @@ func (h *putEndpointID) Handle(params PutEndpointIDParams) middleware.Responder
 				e.RUnlock()
 
-				if hasSidecarProxy {
+				if hasSidecarProxy && e.HasBPFProgram() {
```
Does this need a timeout?

tgraf replied:
Where does it actually block until the BPF program is ready?

ianvernon replied:
This is still WIP. I will mark it as ready for review when it is ready.

ianvernon replied:
To answer your question - the logic in `endpoint.HasBPFProgram()` returns true if the channel is closed, false otherwise (which is achieved via `select`). It shouldn't be blocking. The line you commented on, 274 is within a for loop. If `e.HasBPFProgram()` returns false, the for loop will iterate again. This blocks until the BPF program is ready.

tgraf commented on pkg/endpoint/endpoint.go:
```diff
@@ -694,10 +715,11 @@ func (e *Endpoint) RunK8sCiliumEndpointSync() {
+		Options:       option.NewIntOptions(&EndpointMutableOptionLibrary),
+		Status:        NewEndpointStatus(),
+		state:         state,
+		hasBPFProgram: make(chan struct{}, 0),
```
Is the channel leaked if the endpoint is created but never built?

tgraf commented on daemon/state.go:
```diff
@@ -216,6 +216,8 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) {
 				return
 			}
 
+			ep.CreateBPFProgramChannel()
```
The channel is leaked when the endpoint generation fails and the endpoint is never built again.

ianvernon replied:
By “leaked” do you mean that it persists even if the endpoint is deleted ? Does GC not take effect in that case? If the endpoint is never rebuilt , then its BPF program hasn’t been compiled and the channel is still open.

tgraf replied:
It will persist if anyone is blocking on it. If no references exist anymore, gc will eventually remove it. The current code is all non-blocking but this is a potentially very hard to find memory leak as it assumes that nobody ever blocks on the channel while the endpoint is deleted for a build that has failed.

ianvernon replied:
Added code which closes the channel on endpoint deletion to ensure that the channel is GC'd when the endpoint is removed and that no one is blocking on it. Also removed the exported function to create the channel, and instead set its value internally within the Endpoint package. Let me know if this is what you had in mind.

aanm reviewed (dismissed):
This doesn't have to be a channel as you are not using any features provided by the channel. Since you've similar functions to "set/get" you can use `atomic.StoreInt32` with 0 that represents "no-bpf-program" and 1 that represents "has-bpf-program".

Also, in `daemon/endpoint.go:L251` there's the following comment:
```
// Wait for any successful BPF regeneration, which is indicated by any
// positive policy revision (>0). As long as at least one BPF
// regeneration is successful, the endpoint has network connectivity
// so we can return from the creation API call.
```

So, why do we need to "re-check" there's a bpf program if `revCh` already signalizes that?

tgraf reviewed (commented):
Leaving aside the analysis aside and assuming the race is on the early side. The fix is also incomplete as have not yet written the entry in the lxcmap yet so it does not exist from a routing perspective.

I think we have to do a full BPF plumbing before attempting to configure the proxy for the first time including writing to the lxcmap and syncing the policymap.

Can't we just defer the proxy configuration to the very end and then keep on retrying to build it?

rlenglet committed:
endpoint: Update LXC map before proxy ack wait and signalling

In BPF regeneration, update the LXC map for the endpoint before
waiting for an ACK from Envoy, as the LXC map update is required
for the endpoint to receive any traffic.

Signed-off-by: Romain Lenglet <romain@covalent.io>
Signed-off by: Ian Vernon <ian@cilium.io>

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -727,7 +727,16 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
 
+	// Hook the endpoint into the endpoint table and expose it
+	stats.mapSync.Start()
+	err = lxcmap.WriteEndpoint(epInfoCache)
```
This will introduce drops now as the policy map is not ready when the endpoint is announced.

rlenglet replied:
It is. There are 2 policy map updates. The 1st (and most important one) is done on line 607. The only purpose of the 2nd policy map update below is to add newly created L7 redirects.

tgraf replied:
OK, it still moves the lxcmap update in front of waiting for the ctmap cleanup to finish so we could clean up while initial active connections can start being created.

tgraf replied:
We may actually need to wait for the ct cleanup to complete before the BPF generation is performed

rlenglet replied:
OK, I will add a commit to do that.
I will keep the setting of `Endpoint.ctCleaned` to `true` where it is, i.e. after waiting for proxy ACK, because I don't want to acquire the lock earlier.

tgraf commented on pkg/endpoint/bpf.go:
```diff
@@ -732,7 +732,16 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
 
+	// Hook the endpoint into the endpoint table and expose it
+	stats.mapSync.Start()
+	err = lxcmap.WriteEndpoint(epInfoCache)
```
Before this, the lxcmap update was done last so when the initial endpoint build failed for whatever reason, no entry would exist in the lxcmap and the endpoint couldn't be found. With this, we may leave behind a stale entry in the lxcmap for a program that does not exist. I think we should check whether this is the first build ever, and if so, undo the lxcmap entry by deleting the entry again in the error case to not leave behind an lxcmap entry if no BPF program exists.

tgraf replied:
Discussed offline. This is fine as the program is already kept for egress. An error will be returned to CNI and CNI will remove the endpoint altogether again.

ianvernon replied:
Discussed offline.

If CNI add fails, then the endpoint will be deleted, including its lxcmap entry, so we don't have to clean up if this fails here. The program itself has been installed at this point.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -20,6 +20,7 @@ import (
 	"crypto/md5"
 	"encoding/hex"
 	"fmt"
+	"github.com/cilium/cilium/pkg/maps/lxcmap"
```
imports order

ianvernon replied:
Fixed.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -471,6 +471,8 @@ type Endpoint struct {
 	// is enabled for this endpoint.
 	egressPolicyEnabled bool
 
+	hasBPFProgram chan struct{}
```
Replace with an int32

ianvernon replied:
This is up to developer preference, so I'm not going to change this. The nice thing about having a channel is that if we want to in the future, we can block on the endpoint not having a BPF program installed yet.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -482,6 +484,28 @@ type Endpoint struct {
+// HasBPFProgram returns whether a BPF program has been generated for this
+// endpoint.
+func (e *Endpoint) HasBPFProgram() bool {
+	select {
```
return atomic.LoadInt32(&e.hasBPFProgram) == 1

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -482,6 +484,28 @@ type Endpoint struct {
+// has had its BPF program compiled. If the channel is already closed, this is
+// a no-op.
+func (e *Endpoint) CloseBPFProgramChannel() {
+	select {
```
`atomic.StoreInt32(&e.hasBPFProgram2, 1)`
